### ISSUE TITLE
Implement `norm_sources`

### ISSUE THREAD

**Issue by varun-magesh** [deepspace-ucsb/sdi_pipeline#82]
This function appends an 'refmag' and 'refmagerr' field to a catalog. This field contains a magnitude derived by comparing sources with reference stars from remote databases and calculating a relationship between fluxes in each image and magnitudes from reference stars

```
def norm_sources(hduls, source_ext='CAT', ref_ext='REF', sel_method: ("all", "nrandom", "interactive") = "all", n=None):
    """
    This function appends an 'refmag' and 'refmagerr' field to a catalog.
    These fields contain a magnitude derived by comparing sources with reference stars 
    from remote databases and calculating a relationship between fluxes in each image and 
    magnitudes from reference stars,
    :param hduls: a collection of HDUL to normalize
    :param source_ext: an extname or index of the HDU in HDUL to use as raw source data
    :param ref_ext: an extname or index which specifies an HDU of remote reference star data
    :param sel_method: how to choose the reference stars to use. options are all, nrandom, or
        interactive (which lands you in an interpreter where you can manually select stars)
    :param n: number of refstars to use (only used with "nrandom" option
    :returns: HDULs with 'refmag' and 'refmagerr' fields in the source_ext HDU
    """
```

**Comment by PK0207** [comment]
Transformation between Gaia and PTF spectral data,  and then find the magnitudes by doing scp.optimize I think

**Comment by PK0207** [comment]
Finished writing the part about adding the magnitudes to the CAT table, but it adds a whole new table instead of just overwriting CAT so have to see what to do there. Also, have to learn how to use the Gaia transformation table. PTF uses SDSS g, g-r filters, and gaia transform to that is supposedly not the best but they have errors listed so that's fine. Finally, have to put the whole thing in a for loop, or have to specify that to use the function it needs to be in a for loop if feeding it multiple hduls. 

TLDR; Transform is left, and have to figure out how to overwrite or remove the CAT table.

**Comment by PK0207** [comment]
Method to add mag and magerr for images taken in the SDSS12 g-filter implemented in the `norm_sources` branch. Under file name `PTF_norm.py`

**Comment by PK0207** [comment]
`background_norm.py` pushed up to `norm_sources` branch. Need input from Kelvin as to how to turn this script into a general function as it is heavily reliant on collate.

**Comment by PK0207** [comment]
For the normalization method using Gaia data, sometimes the REF data is shorter than the CAT data. Since REF comes from CAT data, the size of the A-matrix must be determined by the length of REF. Implementing today.

**Comment by PK0207** [comment]
ref.py fixed so there are no Nans, subtract fixed so that CAT isn't over written. PTF_norm.py finds magnitudes for ALL stars in the image.

- Added a bunch of try except ValueError loops so that if ref data and cat data aren't the same length they will still be able to run through. Basically truncated whichever is longer.
- Why is data being dropped? Where is data being dropped? How does one become longer than the other?
- Magnitudes look....a little absurd and the errors are massive. 
- Script writes the data to another hdu, which is appended to the end of the hdul and is not named. I want to change that so that it appends mag and magerr to CAT.

**Comment by PK0207** [comment]
Magnitudes are probably weird because there are a lot of places where there was no data found for g_rp_g, g_bp_g, and ref_mag were not found.

- [x] Make it so that gaia_mag_transformed is only calculated for the sources where all four values: cat_flux, g_rp_g,g_bp_g, and reg_mag are defined (i.e. non-zero)
- [ ] Rename the new hdu CAT, and delete the old CAT
- [x] Also, the except ValueError doesn't catch the difference in lengths for shorter cat_flux. Fix that!

**Comment by PK0207** [comment]
c_best = np.linalg.lstsq(A, cat_flux, rcond=None)[0]
  File "<__array_function__ internals>", line 6, in lstsq
  File "/usr/local/lib/python3.6/dist-packages/numpy/linalg/linalg.py", line 2306, in lstsq
    x, resids, rank, s = gufunc(a, b, rcond, signature=signature, extobj=extobj)

Error shows up when handling the exception. When I run the lines manually on the same data this error doesn't come up so I'm a little confused as to why this is happening.

**Comment by PK0207** [comment]
It was a different error with incorrect dimensions because of the change made to gaia_mag_transformed (to only include values where all three ref data points were present). Fixed a few other dimensionality things and found that there were infinities being introduced (this was all for testing on 62.fits), found that it was from bad ref data (ref mag of 100+) so not super sure what to do there.
Then I changed c_err to not have the np.sum() so that it is no longer infinity. This changes the subsequent fluxes to not have exclusively infinities and nans!

- [x] I want to recheck the various calculations by hand to see if what I have in the code matches up. After that, I am still not sure why the numbers are wonky.
- [x] Plot the cat_flux vs. mag, ref_flux vs. ref_mag (phot_g) - should look roughly logarithmic.

**Comment by PK0207** [comment]
- Cat graph looks like a mess:
![image](https://user-images.githubusercontent.com/30590837/130364026-5cb5bea9-bd58-480f-838e-5a9d480e1812.png)
- Ref graph looks good
![image](https://user-images.githubusercontent.com/30590837/130364058-d6cd55b8-ebde-4632-8181-6bf4b34cb63f.png)

**Comment by PK0207** [comment]
gaia_mag_transformed looks ok (is the gaia data appmag or absmag? absmag because mag zeropoint is applied.) 
In some images, the ref data looks ridiculous. Eg. image 32.fits, phot_bp_mean_mag is of the order 1e-310. Made a fix to only accept values between 0.5 and 25.
[http://slittlefair.staff.shef.ac.uk/teaching/phy217/lectures/principles/L04/index.html](url) <- Vega zero points used by Gaia. Can we apply it too?
- [ ] Determine where absurd values of ref data came from
A and the actual calculation for A seems to differ when the script runs but running the lines myself it seems to be fine.
Flux transformed doesn't look right, looks to small
cat_mag seems normal though? Still getting a significant number of points at mag 20+ which shouldn't be possible but we are calculating absmag rather than appmag, so that might have something to do with it.
After changes made, graph of cat_flux vs. cat_mag looks better:
![image](https://user-images.githubusercontent.com/30590837/130614406-3f366658-d8cd-4ace-9fc2-5ec4c1b07a49.png)
Cat mag is still shorter. Can't re-calculate gaia_mag_transformed because the flux used to normalize and find cat mag is calculated from that mag.
Also, since cat_mag is shorter, not all of the stars have associated magnitudes, and the magnitudes we find are not associated with the right objects.
flux_err is now being calculated as described in the first equation on this page: [SExtractor docs](https://sextractor.readthedocs.io/en/latest/Photom.html)

**Comment by PK0207** [comment]
Testing a way to get all of the mags instead of missing some because of the refs because that doesn't make sense.
What I am doing is curve fitting based on the values found of cat_mag from the reference stars (kind of like calibration but not really).  The function used for curve_fitting is a*log10(flux)+b for just the curve (a rough estimate).
Checking the old mags, new mags, and the goodness of fit visually:
![image](https://user-images.githubusercontent.com/30590837/130636826-278afec9-2941-4a2f-878f-e149edbad1e1.png)
With np.log instead:
![image](https://user-images.githubusercontent.com/30590837/130639183-6c23053c-25c1-4143-be92-3920db1dc9d5.png)
red dots: new magnitudes
blue dots: old magnitudes
blue line: function used with calculated coefficients

Do the np.linalg method of fitting then:
![image](https://user-images.githubusercontent.com/30590837/130648335-a2d8251c-2bd0-41e1-b49b-25caee587e32.png)
red dots: new magnitudes
blue dots: old magnitudes
blue line: function used with calculated coefficients

Not better at all? I am making am mistake in the analysis.
- [x] Change errors based on new mag values.
I have not been ignoring the 0s in mag in my testing so retrying. This does require me to know which flux corresponds to which mag.
Changes made seem to have worked and have recovered all of the mag values (flux v mag graph):
![image](https://user-images.githubusercontent.com/30590837/130654973-30a4c9e4-13bf-4555-8dc2-839091dda671.png)
Now to do error analysis. cat_flux, and cat_mag errors need to be determined.

**Comment by PK0207** [comment]
- [x] Error analysis:

Equation mention in above comment for fluxerr gives LOWER BOUND for error.

- [ ] When writing click command, open images with mode='update', and then:
for i in images:
    norm(i)
    i.flush()
    
 then save and write the hdul.
Read the following:
- [ ] Check for correct calibration ([photometric](http://slittlefair.staff.shef.ac.uk/teaching/phy217/lectures/principles/L04/index.html), Lubin paper, and [Solar extinction](https://kopernio.com/viewer?doi=10.1119%2F1.2192789&token=WzIwODYwODMsIjEwLjExMTkvMS4yMTkyNzg5Il0.qVP_51qxeVcMNz7RXhcdmkDcdyw))
- [ ] [SDSS data release 16](https://www.sdss.org/dr16/)
- [ ] [Finding good reference stars](https://britastro.org/node/8152)
Also, Kelvin is running collate on images in /mnt/annex/pkotta/sditest_output

**Comment by PK0207** [comment]
Changed the way cat_mag empty values are filled in. Only the empty values are filled in by the curve_fit optimization now.

- [ ] Check that optimization function is correct.

Results from phase_diagram.py:
Power graph:
![image](https://user-images.githubusercontent.com/30590837/131199119-1ff84709-80ee-46ea-9e95-c51cccaac0fb.png)
Phase diagram:
![image](https://user-images.githubusercontent.com/30590837/131199212-2063b63f-db9a-4a21-9687-0856ce937ec4.png)

- [x] Does the time for a datapoint match up to when that data point was actually recorded?

**Comment by PK0207** [comment]
Sorted times in increasing order because that's how the images are arranged, and therefore how the mags are arranged. The periodogram looked like that because it was such a minute frequency window we were looking at. For a larger window it looks fine:
![image](https://user-images.githubusercontent.com/30590837/131226167-1ab3cc51-cca5-4f78-a8e6-84ce5216d4d7.png)
Improvement: clear peak, less noise because of the better magnitude errors
But the peak is still not very large (0.175 out of 1), again possibly because of errors.
Light curve (with magerrs found by periodogram):
![image](https://user-images.githubusercontent.com/30590837/131226494-b8c1a291-5960-4dd0-b333-16111dcf8832.png)

Light curve (with magerrs, with known period used with phase folding function):
![image](https://user-images.githubusercontent.com/30590837/131226566-9434e49b-985f-46ad-86b0-66cf87034102.png)
Clear periodicity but still not clear enough.
Errors look huge: +/- 20 mag???

**Comment by PK0207** [comment]
Times don't match up exactly between PTF times and times from the hduls.

- [ ] Make sure that reference stars found are not on the edge. Maybe put this in ref.py

Try with 1 reference star (random number 37)
Not all images have 37 reference stars: 93, 62,43,30 <- normalized manually with ref[30]. Did not check which star each one was!

Plotted a histogram of differences between calculated magnitudes and the transformed ref magnitudes.:
![image](https://user-images.githubusercontent.com/30590837/131678890-9c40652b-8271-47be-983f-0579acc9e08b.png)
This shows very large differences in mag (preferably the differences would be less than 1 but the min and max go from -4 to 8).
This is why we see so much more spread in our lightcurve than in the PTF one.
Magnitude errors might account for these differences in some part (max error is +- 4 which could account for differences in mag up to 8):
![image](https://user-images.githubusercontent.com/30590837/131679675-bd4a8aef-874a-4650-90b6-86772960dcd7.png)
Actually, they almost certainly are because of the spread in |diffs/2|:
![image](https://user-images.githubusercontent.com/30590837/131680226-cb20a5e6-23b9-4d89-af8b-5e33149ce10d.png)

**Comment by PK0207** [comment]
![image](https://user-images.githubusercontent.com/30590837/133104749-06e2743f-028f-4eb0-8b15-c8222d4837ee.png)
Red: All sources
Green: Reference stars (converted RADEC to xy
Blue: Variable stars

Ref stars are not circled properly. I don't think the conv RD_XY code works.

**Comment by PK0207** [comment]
- [x] Threshold value needs to be bigger because images are so small

**Comment by PK0207** [comment]
Choosing reference stars issue has been resolved, and methodology has been changed to using a linear fit on the reference stars and using that to correct the target star magnitude. This is a simple photometric process. Now we need to ensure that the program is tested, and is written more efficiently. Currently, the program uses a lot of memory. Also, the click command norm_sources is not implemented. This needs some testing of collate and the performance of subtract. Then we can generalize the code to perform normalization for all the variable stars in the image at once.

**Comment by PK0207** [comment]
TODO:

- [ ]  Averaging method for corrections
- [ ] Make a function to find common coordinates between lists
- [ ] Make script memory efficient
- [ ] Make it so that we can run multiple target stars through at once
- [ ] Use a better variable for aperture in photometry
- [ ] Normalize ref_mags outside the for loop
- [ ] Run the script on the variable stars that the pipeline finds
- [ ] Make it so that we can easily run one reference star through
- [ ] Color matching

**Comment by PK0207** [comment]
![image](https://user-images.githubusercontent.com/30590837/156093138-5d848ed2-f811-47d6-b61a-e5b3b25f0f43.png)
This is the resulting  lightcurve using PTF images  run through our pipeline. Red indicates the magnitudes calculated by PTF.

See here for more plots:
https://docs.google.com/presentation/d/1jWwEcMXRi3JeFrHxzDnDg347n7Dk1SO9RWHnHOPvS7c/edit?usp=sharing
